File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.scala
import scala.collection.mutable

object Solution {
    def maxSum(nums: Array[Int], k: Int): Long = {

        val n = nums.length

        var ans = Int.MinValue.toLong

        if (k == 0 || n == 1) {
            var sm = 0L

            for (num <- nums) {
                sm += num
                ans = math.max(ans, sm)

                if (sm < 0)
                    sm = 0
            }

            return ans
        }

        var cnt = 0
        var cur = 0L
        val pref = Array.fill[Long](n + 1)(0L)

        for (i <- 0 until n) {
            if (nums(i) >= 0)
                cur += nums(i)
            else
                cnt += 1

            pref(i + 1) = pref(i) + nums(i)
            ans = math.max(ans, nums(i).toLong)
        }

        if (cnt <= k)
            return cur

        val dp = Array.fill[Long](n, n)(0L)

        for (i <- 0 until n) {
            val pq = mutable.PriorityQueue[Int]()
            var sm = 0L

            for (j <- i until n) {
                if (nums(j) >= 0) {
                    dp(i)(j) = sm
                } else {
                    if (pq.size < k) {
                        pq.enqueue(nums(j))
                        sm += nums(j)
                    } else if (pq.head > nums(j)) {
                        sm -= pq.dequeue()
                        pq.enqueue(nums(j))
                        sm += nums(j)
                    }
                }
            }
        }

        for (i <- 0 until n) {
            implicit val ord: Ordering[Int] = Ordering.Int.reverse
            val pqMax = mutable.PriorityQueue.empty[Int]
            var sm = 0L

            for (j <- 0 until i) {
                if (nums(j) >= 0) {
                    if (pqMax.size < k) {
                        pqMax.enqueue(nums(j))
                        sm += nums(j)
                    } else if (pqMax.head < nums(j)) {
                        sm -= pqMax.dequeue()
                        pqMax.enqueue(nums(j))
                        sm += nums(j)
                    }
                }
            }

            for (j <- (i + 1 until n).reverse) {
                var curSum = pref(j + 1) - pref(i)
                curSum -= dp(i)(j)
                curSum += sm

                ans = math.max(ans, curSum)

                if (nums(j) >= 0) {
                    if (pqMax.size < k) {
                        pqMax.enqueue(nums(j))
                        sm += nums(j)
                    } else if (pqMax.head < nums(j)) {
                        sm -= pqMax.dequeue()
                        pqMax.enqueue(nums(j))
                        sm += nums(j)
                    }

                    if (pqMax.nonEmpty)
                        ans = math.max(ans, sm)
                }
            }
        }

        ans
    }
}

File: 3962-maximum-subarray-sum-after-at-most-k-swaps/3962-maximum-subarray-sum-after-at-most-k-swaps.js
var maxSum = function(nums, k) {
    const n = nums.length;

    let ans = -2147483648;

    if (k === 0 || n === 1) {
        let sm = 0;
        for (const num of nums) {
            sm += num;
            ans = Math.max(ans, sm);
            if (sm < 0) sm = 0;
        }
        return ans;
    }

    let cnt = 0;
    let cur = 0;
    const pref = new Array(n + 1).fill(0);

    for (let i = 0; i < n; i++) {
        if (nums[i] >= 0) cur += nums[i];
        else cnt++;

        pref[i + 1] = pref[i] + nums[i];
        ans = Math.max(ans, nums[i]);
    }

    if (cnt <= k) return cur;

    const dp = Array.from({ length: n }, () => Array(n).fill(0));

    for (let i = 0; i < n; i++) {
        const pq = [];
        let sm = 0;

        for (let j = i; j < n; j++) {
            if (nums[j] >= 0) {
                dp[i][j] = sm;
                continue;
            }

            if (pq.length < k) {
                pq.push(nums[j]);
                pq.sort((a, b) => b - a);
                sm += nums[j];
            } else if (pq[0] > nums[j]) {
                sm -= pq[0];
                pq[0] = nums[j];
                pq.sort((a, b) => b - a);
                sm += nums[j];
            }
        }
    }

    for (let i = 0; i < n; i++) {
        const pqmax = [];
        let sm = 0;

        for (let j = 0; j < i; j++) {
            if (nums[j] < 0) continue;

            if (pqmax.length < k) {
                pqmax.push(nums[j]);
                pqmax.sort((a, b) => a - b);
                sm += nums[j];
            } else if (pqmax[0] < nums[j]) {
                sm -= pqmax[0];
                pqmax[0] = nums[j];
                pqmax.sort((a, b) => a - b);
                sm += nums[j];
            }
        }

        for (let j = n - 1; j > i; j--) {
            let cur = pref[j + 1] - pref[i];
            cur -= dp[i][j];
            cur += sm;

            ans = Math.max(ans, cur);

            if (nums[j] < 0) continue;

            if (pqmax.length < k) {
                pqmax.push(nums[j]);
                pqmax.sort((a, b) => a - b);
                sm += nums[j];
            } else if (pqmax[0] < nums[j]) {
                sm -= pqmax[0];
                pqmax[0] = nums[j];
                pqmax.sort((a, b) => a - b);
                sm += nums[j];
            }

            if (pqmax.length > 0) {
                ans = Math.max(ans, sm);
            }
        }
    }

    return ans;
};